src/Microsoft.TemplateEngine.TemplateLocalizer.Core/KeyExtractors/IJsonKeyCreator.cs

```diff
@@ -0,0 +1,24 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.Json;
+
+namespace Microsoft.TemplateEngine.TemplateLocalizer.Core.KeyExtractors
+{
+    /// <summary>
+    /// Creates keys for a given <see cref="JsonElement"/> of a JSON document that can be used to uniquely identify the element
+    /// among its siblings under the same parent <see cref="JsonElement"/>.
+    /// </summary>
+    internal interface IJsonKeyCreator
+    {
+        /// <summary>
+        /// Creates a key for the given <see cref="JsonElement"/>.
+        /// </summary>
+        /// <param name="element">The element that the key will uniquely represent under the same parent.</param>
+        /// <param name="elementName">Name of the element that the key is being created for.</param>
+        /// <param name="parentElementName">Name of the parent element.</param>
+        /// <param name="indexInParent">Index of this element with respect to its siblings under the same parent.</param>
+        /// <returns>The key string.</returns>
+        string CreateKey(JsonElement element, string? elementName, string? parentElementName, int indexInParent);
```

Comment:
I think I get the intention of this interface now, but thinking whether this can be just an extension of JSON object or `Identifier` string instead. In theory, for each JSON element it should be possible to traverse up to the parent and generate `Identifier` element when needed, and not preserving the identifier/key name when traversing.  Probably this can make the code and recursion logic easier.

Reply:
That would work for identifiers where we always expect the same behavior regardless of the json element we are processing. However, `IJsonKeyCreator` is for key generation, not identifiers. Keys can be different in depending on which part of the template.json the related json element occurs.

Reply:
As far as I see identifiers are still manually concatenated on recursion

Reply:
Yes. I was trying to explain why we can't remove `IJsonKeyCreator`, which is about keys and not identifiers. Since there is another discussion comment in the PR about identifiers, I'm marking this one resolved. Please unresolve if this is still relevant to the discussion.